### PR TITLE
new gensym handling

### PR DESCRIPTION
**Pull request by Araq**
- will fix many template bugs, eventually.

### CODE REVIEW

File: doc/manual.rst [hunk 1]
```diff
@@ -4893,6 +4893,8 @@ no semantics outside of a template definition and cannot be abstracted over:
 To get rid of hygiene in templates, one can use the `dirty`:idx: pragma for
 a template. ``inject`` and ``gensym`` have no effect in ``dirty`` templates.
 
+``gensym``'ed symbols cannot be used
```
Comment (juancarlospaco) [inline doc/manual.rst]:
```suggestion
``gensym`` processed symbols cannot be used.
```

Yo 'cause slang kinda hard 2 read 4 n00bz. 🤪

Comment (Araq) [reply doc/manual.rst]: It's work-in-progress to remind me to write this part of the spec.

File: compiler/commands.nim [hunk 2]
```diff
@@ -788,6 +788,15 @@ proc processSwitch*(switch, arg: string, pass: TCmdLinePass, info: TLineInfo;
   of "expandmacro":
     expectArg(conf, switch, arg, pass, info)
     conf.macrosToExpand[arg] = "T"
+  of "useversion":
+    expectArg(conf, switch, arg, pass, info)
+    case arg
+    of "0.19":
+      conf.globalOptions.incl optNimV019
```
Comment (timotheecour) [inline compiler/commands.nim]: nit: is `optNimV019` related to `nimv019` in config/nim.cfg? as i understand this only affects gensym handling, so maybe it should be in the flag name, eg `--genSymVersion`

Comment (Araq) [reply compiler/commands.nim]: We are moving over to "trunk based" development, that's why. These transition switches will be collected under ``useVersion``.

**Comment by timotheecour** [comment]
I thought this PR would fix following bugs but apparently it doesn't, and maybe it can't as there's inherent ambiguity:

```nim
type Foo = object
  b: int
proc bar(b = 12) = discard

template fun(b): untyped =
  var x: Foo
  # each line here would give an error
  var x0 = Foo(b: 3)
  bar(b=32)
  x.b = 5
  echo x.b

template fun2(b2): untyped =
  var x = Foo(b: 3)
  x.b = 5
  echo x.b
  echo b2

proc main()=
  fun2(14) # ok
  fun(14) # Error: identifier expected, but found '14'

main()

```

IMO we should reopen https://github.com/nim-lang/Nim/issues/8899, it's a fool proof solution:
```nim
template fun(b): untyped =
  var x: Foo
  var x0 = Foo(`b`: 3)
  bar(`b`=32)
  x.`b` = 5
  x.b = 5 # this one actually means: substitute (eg if user passes an identifier as b, it will work as intended)
  echo x.`b`
```

**Pull request merged**